presto-main/src/test/java/com/facebook/presto/type/TestTimeBase.java

```diff
@@ -174,6 +179,36 @@ public void testCastToTimeWithTimeZone()
                 new SqlTimeWithTimeZone(new DateTime(1970, 1, 1, 3, 4, 5, 321, DATE_TIME_ZONE).getMillis(), DATE_TIME_ZONE.toTimeZone()));
     }
 
+    @Test
+    public void testCastToTimeWithTimeZoneWithTZWithRulesChanged()
+    {
+        TimeZoneKey timeZoneThatChangedSince1970 = getTimeZoneKey("Asia/Kathmandu");
+        DateTimeZone dateTimeZoneThatChangedSince1970 = getDateTimeZone(timeZoneThatChangedSince1970);
+        Session session = testSessionBuilder()
+                .setTimeZoneKey(timeZoneThatChangedSince1970)
+                .build();
+        FunctionAssertions localAssertions = new FunctionAssertions(session);
+
+        localAssertions.assertFunction(
+                "cast(TIME '03:04:05.321' as time with time zone)",
+                TIME_WITH_TIME_ZONE,
+                new SqlTimeWithTimeZone(new DateTime(1970, 1, 1, 3, 4, 5, 321, dateTimeZoneThatChangedSince1970).getMillis(), dateTimeZoneThatChangedSince1970.toTimeZone()));
+    }
+
+    @Test
+    public void testCastToTimestampWithTimeZoneDSTIsNotAppliedWhenTimeCrossesDST()
+    {
+        // Australia/Sydney will switch DST a second after session start
+        // For simplicity we have to use time zone that is going forward when entering DST zone with 1970-01-01
+        Session session = testSessionBuilder()
+                .setTimeZoneKey(getTimeZoneKey("Australia/Sydney"))
+                .setStartTime(new DateTime(2017, 10, 1, 1, 59, 59, 999, getDateTimeZone(getTimeZoneKey("Australia/Sydney"))).getMillis())
+                .build();
+        FunctionAssertions localAssertions = new FunctionAssertions(session);
```

Comment:
`FunctionsAssertions` are closeable (and contain `LocalQueryRunner`) and should be closed.